privacyIDEAADFSProvider/Adapter.cs

```diff
@@ -210,64 +220,71 @@ public IAuthenticationAdapterMetadata Metadata
-                        response = privacyIDEA.ValidateCheck(user, "", transactionid);
-                        if (response != null && response.Value)
-                        {
-                            outgoingClaims = Claims();
-                            return null;
-                        }
-                    }
-                    else
-                    {
-                        // Else push not confirmed yet
-                        form.ErrorMessage = "Authenication not confirmed yet!";
-                    }
+                    // Push confirmed, finish the authentication via /validate/check using an empty otp
+                    // https://privacyidea.readthedocs.io/en/latest/tokens/authentication_modes.html#outofband-mode
+                    response = privacyIDEA.ValidateCheck(user, "", transactionid);
                 }
-                else if (mode == "webauthn")
+                else
                 {
-                    // TODO webauthn implementaiton
+                    // Else push not confirmed yet
+                    form.ErrorMessage = "Authenication not confirmed yet!";
+                }
+            }
+            else if (mode == "webauthn")
+            {
+                string origin = (string)GetFromDict(proofDict, "origin");
+                string webauthnresponse = (string)GetFromDict(proofDict, "webAuthnSignResponse");
+
+                if (string.IsNullOrEmpty(origin) || string.IsNullOrEmpty(webauthnresponse))
+                {
+                    Error("Incomplete data for WebAuthn authentication: WebAuthnSignResponse or Origin is missing!");
+                    form.ErrorMessage = "Could not complete WebAuthn authentication. Try again or use another token type.";
                 }
                 else
                 {
-                    // Mode == OTP
-                    response = privacyIDEA.ValidateCheck(user, otp, transactionid);
-                    if (response != null && response.Value)
-                    {
-                        outgoingClaims = Claims();
-                        return null;
-                    }
+                    response = privacyIDEA.ValidateCheckWebAuthn(user, transactionid, webauthnresponse, origin);
                 }
+            }
+            else
+            {
+                // Mode == OTP
+                response = privacyIDEA.ValidateCheck(user, otp, transactionid);
+            }
 
-                // If we get this far, the login data provided was wrong or an error occured.
-                if (response != null)
+            // If we get this far, the login data provided was wrong, an error occured or another challenge was triggered.
+            if (response != null)
+            {
+                if (response.MultiChallenge.Count > 0)
+                {
+                    // TODO 
+                    Log("multichallenge > 0");
```

Comment:
What do you mean with "multichallenge" here?

I would like to get rid of the term "multi_challenge" in the long run, so maybe you can already talk of `response.challenges.Count`
or is the object created automatically from the key names like

     multi_challenge -> MultiChallenge

?

Reply:
In this case it is the name of the field in the PIResponse class, so it could already be renamed.

Reply:
It is up to you how you (or others) will understand the code better.